radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java

```diff
@@ -732,12 +760,64 @@ private void executeStateUpdate(com.sleepycat.je.Transaction txn, REStateUpdate
 				var buf2 = stateUpdate.getStateBuf();
 				var value = new DatabaseEntry(buf2.array(), buf2.position(), buf2.remaining());
 				resourceDatabase.putNoOverwrite(txn, new DatabaseEntry(addr.getBytes()), value);
+			} else if (stateUpdate.getParsed() instanceof VirtualParent) {
+				var p = (VirtualParent) stateUpdate.getParsed();
+				var typeByte = p.getData()[0];
+				if (typeByte != SubstateTypeId.UNCLAIMED_READDR.id()) {
+					var mapKey = SystemMapKey.ofValidatorDataParent(typeByte);
+					var key = new DatabaseEntry(mapKey.array());
+					var value = new DatabaseEntry(stateUpdate.getId().asBytes());
+					var result = mapDatabase.putNoOverwrite(txn, key, value);
+					if (result != SUCCESS) {
+						throw new IllegalStateException();
+					}
+				}
+			} else if (stateUpdate.getParsed() instanceof ValidatorData) {
+				var p = (ValidatorData) stateUpdate.getParsed();
+				var mapKey = SystemMapKey.ofValidatorData(
+					stateUpdate.typeByte(),
+					p.getValidatorKey().getCompressedBytes()
+				);
+				var key = new DatabaseEntry(mapKey.array());
+				var value = new DatabaseEntry(stateUpdate.getId().asBytes());
+				var result = mapDatabase.putNoOverwrite(txn, key, value);
+				if (result != SUCCESS) {
+					throw new IllegalStateException();
+				}
+			} else if (stateUpdate.getParsed() instanceof SystemData) {
+				var mapKey = SystemMapKey.ofSystem(stateUpdate.typeByte());
+				var key = new DatabaseEntry(mapKey.array());
+				var value = new DatabaseEntry(stateUpdate.getId().asBytes());
+				var result = mapDatabase.putNoOverwrite(txn, key, value);
+				if (result != SUCCESS) {
+					throw new IllegalStateException();
+				}
 			}
 		} else if (stateUpdate.isShutDown()) {
 			if (stateUpdate.getId().isVirtual()) {
 				downVirtualSubstate(txn, stateUpdate.getId());
 			} else {
 				downSubstate(txn, stateUpdate.getId());
+
+				if (stateUpdate.getParsed() instanceof ValidatorData) {
+					var p = (ValidatorData) stateUpdate.getParsed();
+					var mapKey = SystemMapKey.ofValidatorData(
+						stateUpdate.typeByte(),
+						p.getValidatorKey().getCompressedBytes()
+					);
+					var key = new DatabaseEntry(mapKey.array());
+					var result = mapDatabase.delete(txn, key);
+					if (result != SUCCESS) {
+						throw new IllegalStateException();
+					}
```

Comment:
Similar to above case - logic repeated few times.

Reply:
good point